graph/appender/dead_node.go

```diff
@@ -0,0 +1,84 @@
+package appender
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/kiali/kiali/graph"
+	"github.com/kiali/kiali/kubernetes"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// DeadNodeAppender is responsible for removing from the graph any nodes for which
+// there is no traffic reported and the related definitions are undefined (presumably
+// removed from K8S) and for which . (kiali-621)
+type DeadNodeAppender struct{}
+
+// AppendGraph implements Appender
+func (a DeadNodeAppender) AppendGraph(trafficMap graph.TrafficMap, _ string) {
+	if len(trafficMap) == 0 {
+		return
+	}
+
+	istioClient, err := kubernetes.NewClient()
+	checkError(err)
+
+	applyDeadNodes(trafficMap, istioClient)
+}
+
+func applyDeadNodes(trafficMap graph.TrafficMap, istioClient kubernetes.IstioClientInterface) {
+	numRemoved := 0
+	for id, n := range trafficMap {
+		// a node with traffic is not dead, skip
+		rate, hasRate := n.Metadata["rate"]
+		rateOut, hasRateOut := n.Metadata["rateOut"]
+		if (hasRate && rate.(float64) > 0) || (hasRateOut && rateOut.(float64) > 0) {
+			continue
+		}
+		// Remove if backing deployment is not defined, flag if there are no pods
+		// Note that in the future a workload may not fron only a deployment, in that case
+		// the workload name may need to be queried against multiple potential things.
+		deployment, err := istioClient.GetDeployment(n.Namespace, n.Workload)
+		if err != nil || deployment == nil {
+			delete(trafficMap, id)
+			numRemoved++
+		} else {
+			pods, err := istioClient.GetPods(n.Namespace, selectorAsString(deployment.Spec.Selector))
+			if err != nil || pods == nil || len(pods.Items) == 0 {
+				n.Metadata["isDead"] = true
+			}
+		}
+	}
+
+	// If we removed any nodes we need to remove any edges to them as well...
+	if numRemoved == 0 {
+		return
+	}
+
+	for _, s := range trafficMap {
+		goodEdges := []*graph.Edge{}
+		for _, e := range s.Edges {
+			if _, found := trafficMap[e.Dest.ID]; found {
+				goodEdges = append(goodEdges, e)
+			}
+		}
+		s.Edges = goodEdges
+	}
+}
+
+func selectorAsString(selector *meta_v1.LabelSelector) string {
```

Comment:
I'll try it but I because I didn't see a String imp for LabelSelector I thought it wouldn't work.  Let me try and if it works I'll convert.  I suppose I should have tried because that was my first thought....

Reply:
Have a look at kubernetes_services.go line 149, we already do it.

Reply:
Great, works perfectly.   I knew there had to be support for handling their own labels structure, I just couldn't find it...  Updating...